File: examples/python/ImageAlign/custom_message_align.py
#!/usr/bin/env python3

import cv2
import depthai as dai

FPS = 30.0
SourceSize = (640, 640)
AlignToSize = (1280, 720)

class Line(dai.TransformableBuffer):
    def __init__(self):
        super().__init__()
        self.startPoint: dai.Point2f = None
        self.endPoint: dai.Point2f = None

    # Need to override the base method to properly transform custom messages
    def transformTo(self, target: dai.ImgTransformation) -> dai.TransformableBuffer:
        source = self.getTransformation()
        assert source is not None, "Source transformation is not set for the line"

        out = Line()
        out.startPoint = source.remapPointTo(target, self.startPoint)
        out.endPoint = source.remapPointTo(target, self.endPoint)
        out.setTransformation(target)
        return out


def drawLine(frame, line: Line, color, label: str) -> None:
    start = (int(round(line.startPoint.x)), int(round(line.startPoint.y)))
    end = (int(round(line.endPoint.x)), int(round(line.endPoint.y)))

    cv2.line(frame, start, end, color, 1)
    cv2.circle(frame, start, 2, color, -1)
    cv2.circle(frame, end, 2, color, -1)
    cv2.putText(frame, label, (start[0] + 12, max(24, start[1] - 12)), cv2.FONT_HERSHEY_TRIPLEX, 0.6, color)


def makeLine(SourceFrame: dai.ImgFrame) -> Line:
    line = Line()
    line.setTransformation(SourceFrame.getTransformation())
    line.startPoint = dai.Point2f(110.0, 150.0)
    line.endPoint = dai.Point2f(520.0, 430.0)
    return line


if __name__ == "__main__":
    pipeline = dai.Pipeline()
    camera = pipeline.create(dai.node.Camera).build(boardSocket=dai.CameraBoardSocket.CAM_A, sensorFps=FPS)

    sourceOutput = camera.requestOutput(size=SourceSize, fps=FPS, resizeMode=dai.ImgResizeMode.LETTERBOX, enableUndistortion=False)
    alignToOutput = camera.requestOutput(size=AlignToSize, fps=FPS, resizeMode=dai.ImgResizeMode.STRETCH, enableUndistortion=True)

    align = pipeline.create(dai.node.ImageAlign)
    align.setRunOnHost(True) # for custom message, ImageAlign needs to run on host.
    alignToOutput.link(align.inputAlignTo)

    sourceQueue = sourceOutput.createOutputQueue()
    alignToQueue = alignToOutput.createOutputQueue()
    alignedLineQueue = align.outputAligned.createOutputQueue()

    lineInputQueue = align.input.createInputQueue()

    pipeline.start()

    while pipeline.isRunning():
        sourceFrameMsg: dai.ImgFrame = sourceQueue.get()
        alignToFrameMsg: dai.ImgFrame = alignToQueue.get()

        sourceLine = makeLine(sourceFrameMsg)
        lineInputQueue.send(sourceLine)
        alignedLine: Line = alignedLineQueue.get()
        assert isinstance(alignedLine, Line)

        sourceFrame = sourceFrameMsg.getCvFrame()
        alignToFrame = alignToFrameMsg.getCvFrame()

        drawLine(sourceFrame, sourceLine, (0, 255, 0), "source line")
        drawLine(alignToFrame, alignedLine, (0, 140, 255), "aligned line")

        cv2.putText(sourceFrame, "source: CROP 640x640", (10, 28), cv2.FONT_HERSHEY_TRIPLEX, 0.6, (255, 255, 255))
        cv2.putText(alignToFrame, "alignTo: STRETCH 1280x720", (10, 28), cv2.FONT_HERSHEY_TRIPLEX, 0.6, (255, 255, 255))

        cv2.imshow("Source output", sourceFrame)
        cv2.imshow("Aligned output", alignToFrame)

        if cv2.waitKey(1) == ord("q"):
            pipeline.stop()
            break

File: examples/cpp/Align/custom_message_align.cpp
#include <algorithm>
#include <atomic>
#include <cmath>
#include <csignal>
#include <memory>
#include <opencv2/opencv.hpp>
#include <stdexcept>
#include <string>

#include "depthai/depthai.hpp"

constexpr float FPS = 30.0f;
const std::pair<uint32_t, uint32_t> SOURCE_SIZE{640, 640};
const std::pair<uint32_t, uint32_t> ALIGN_TO_SIZE{1280, 720};

std::atomic<bool> quitEvent(false);

void signalHandler(int) {
    quitEvent = true;
}

// Custom message carrying a line segment. Overriding transformTo() lets the Align node remap it.
class Line : public dai::TransformableBuffer {
   public:
    dai::Point2f startPoint;
    dai::Point2f endPoint;

    std::shared_ptr<dai::TransformableBuffer> transformTo(const dai::ImgTransformation& target) const override {
        auto source = getTransformation();
        if(!source.has_value()) {
            throw std::runtime_error("Source transformation is not set for the line");
        }

        auto out = std::make_shared<Line>();
        out->startPoint = source->remapPointTo(target, startPoint);
        out->endPoint = source->remapPointTo(target, endPoint);
        out->setTransformation(target);
        return out;
    }
};

void drawLine(cv::Mat& frame, const Line& line, const cv::Scalar& color, const std::string& label) {
    const cv::Point start(static_cast<int>(std::lround(line.startPoint.x)), static_cast<int>(std::lround(line.startPoint.y)));
    const cv::Point end(static_cast<int>(std::lround(line.endPoint.x)), static_cast<int>(std::lround(line.endPoint.y)));

    cv::line(frame, start, end, color, 1);
    cv::circle(frame, start, 2, color, -1);
    cv::circle(frame, end, 2, color, -1);
    cv::putText(frame, label, {start.x + 12, std::max(24, start.y - 12)}, cv::FONT_HERSHEY_TRIPLEX, 0.6, color);
}

std::shared_ptr<Line> makeLine(const dai::ImgFrame& sourceFrame) {
    auto line = std::make_shared<Line>();
    line->setTransformation(sourceFrame.transformation);
    line->startPoint = dai::Point2f(110.0f, 150.0f);
    line->endPoint = dai::Point2f(520.0f, 430.0f);
    return line;
}

int main() {
    signal(SIGTERM, signalHandler);
    signal(SIGINT, signalHandler);

    dai::Pipeline pipeline;
    auto camera = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_A, std::nullopt, FPS);

    auto* sourceOutput = camera->requestOutput(SOURCE_SIZE, std::nullopt, dai::ImgResizeMode::LETTERBOX, FPS, false);
    auto* alignToOutput = camera->requestOutput(ALIGN_TO_SIZE, std::nullopt, dai::ImgResizeMode::STRETCH, FPS, true);

    auto align = pipeline.create<dai::node::Align>();
    align->setRunOnHost(true);  // for custom messages, Align needs to run on host.
    alignToOutput->link(align->inputAlignTo);

    auto sourceQueue = sourceOutput->createOutputQueue();
    auto alignToQueue = alignToOutput->createOutputQueue();
    auto alignedLineQueue = align->outputAligned.createOutputQueue();

    auto lineInputQueue = align->input.createInputQueue();

    pipeline.start();

    while(pipeline.isRunning() && !quitEvent) {
        auto sourceFrameMsg = sourceQueue->get<dai::ImgFrame>();
        auto alignToFrameMsg = alignToQueue->get<dai::ImgFrame>();
        if(sourceFrameMsg == nullptr || alignToFrameMsg == nullptr) {
            continue;
        }

        auto sourceLine = makeLine(*sourceFrameMsg);
        lineInputQueue->send(sourceLine);
        auto alignedLine = alignedLineQueue->get<Line>();
        if(alignedLine == nullptr) {
            throw std::runtime_error("Aligned message is not a Line");
        }

        cv::Mat sourceFrame = sourceFrameMsg->getCvFrame();
        cv::Mat alignToFrame = alignToFrameMsg->getCvFrame();

        drawLine(sourceFrame, *sourceLine, {0, 255, 0}, "source line");
        drawLine(alignToFrame, *alignedLine, {0, 140, 255}, "aligned line");

        cv::putText(sourceFrame, "source: LETTERBOX 640x640", {10, 28}, cv::FONT_HERSHEY_TRIPLEX, 0.6, {255, 255, 255});
        cv::putText(alignToFrame, "alignTo: STRETCH 1280x720", {10, 28}, cv::FONT_HERSHEY_TRIPLEX, 0.6, {255, 255, 255});

        cv::imshow("Source output", sourceFrame);
        cv::imshow("Aligned output", alignToFrame);

        if(cv::waitKey(1) == 'q') {
            pipeline.stop();
            break;
        }
    }

    cv::destroyAllWindows();
    return 0;
}
